app/js/services/auth.js

```diff
@@ -0,0 +1,82 @@
+import Cookies from 'js-cookie'
+import axios from 'axios'
+import { Promise } from 'es6-promise'
+import { CP_ADMIN_HOST, CP_ANALYTICS_HOST } from '../constants/app_constants'
+import { routeUtils } from '../helpers/router'
+
+const LOGIN_URL = CP_ADMIN_HOST + '#/auth/sign_in?return_url=' + CP_ANALYTICS_HOST
+const SESSION_PING_URL = CP_ADMIN_HOST + 'api/v1/session'
+const REFRESH_TOKEN_URL = CP_ADMIN_HOST + 'api/v1/api_session/refresh'
+
+const Auth = {
+  init () {
+    this.addXHRInterceptor()
+    this.initSession()
+  },
+
+  addXHRInterceptor () {
+    axios.interceptors.request.use(
+      this.addXHRHeaders.bind(this),
+      (error) => { return Promise.reject(error) }
+    )
+
+    axios.interceptors.response.use(
+      (response) => { return response },
+      (error) => { return this.resolveRejectedRequest(error) }
+    )
+  },
+
+  initSession () {
+    axios.get(SESSION_PING_URL)
+  },
+
+  addXHRHeaders (request) {
+    request.headers['Authorization'] = this.getAccessToken()
+    return request
+  },
+
+  resolveRejectedRequest (response) {
+    if (response.status === 401) {
+      return this.refreshToken(response).then((request) => {
+        // Restart previous request after token is refreshed
+        this.setAccessToken(request.data.access_token)
+        this.restartRequest(response)
+      }).catch(this.redirectToLoginPage)
+    } else {
+      return Promise.reject(response)
+    }
+  },
+
+  refreshToken (response) {
+    return axios.post(REFRESH_TOKEN_URL, {
+      refresh_token: this.getRefreshToken(),
+    })
+  },
+
+  redirectToLoginPage () {
+    routeUtils.redirectTo(LOGIN_URL)
+  },
+
+  getAccessToken () {
+    return Cookies.get('accessToken')
+  },
+
+  setAccessToken (newToken) {
+    Cookies.set('accessToken', newToken)
+  },
+
+  getRefreshToken () {
+    return Cookies.get('refreshToken')
+  },
+
+  restartRequest (response) {
+    return axios({
+      method: response.config.method,
+      url: response.config.url,
+      data: response.data,
+    })
+  },
+
+}
+
+export default Auth
```

Comment:
Ahh... that is perfect. Thank you very much man. I will change the code regarding to your advice.